# Logistic Regression Consulting Project

## Binary Customer Churn

A marketing agency has many customers that use their service to produce ads for the client/customer websites. They've noticed that they have quite a bit of churn in clients. They basically randomly assign account managers right now, but want you to create a machine learning model that will help predict which customers will churn (stop buying their service) so that they can correctly assign the customers most at risk to churn an account manager. Luckily they have some historical data, can you help them out? Create a classification algorithm that will help classify whether or not a customer churned. Then the company can test this against incoming data for future customers to predict which customers will churn and assign them an account manager.

The data is saved as customer_churn.csv. Here are the fields and their definitions:

    Name : Name of the latest contact at Company
    Age: Customer Age
    Total_Purchase: Total Ads Purchased
    Account_Manager: Binary 0=No manager, 1= Account manager assigned
    Years: Totaly Years as a customer
    Num_sites: Number of websites that use the service.
    Onboard_date: Date that the name of the latest contact was onboarded
    Location: Client HQ Address
    Company: Name of Client Company
    
Once you've created the model and evaluated it, test out the model on some new data (you can think of this almost like a hold-out set) that your client has provided, saved under new_customers.csv. The client wants to know which customers are most likely to churn given this data (they don't have the label yet).

In [1]:
import findspark 
findspark.init('/home/gkouskosv/spark-2.4.5-bin-hadoop2.6/')

In [2]:
from pyspark.sql import SparkSession

spark = SparkSession.builder.appName('ConProj').getOrCreate()

In [3]:
data = spark.read.csv('customer_churn.csv',header=True,inferSchema=True)

In [4]:
data.printSchema()

root
 |-- Names: string (nullable = true)
 |-- Age: double (nullable = true)
 |-- Total_Purchase: double (nullable = true)
 |-- Account_Manager: integer (nullable = true)
 |-- Years: double (nullable = true)
 |-- Num_Sites: double (nullable = true)
 |-- Onboard_date: timestamp (nullable = true)
 |-- Location: string (nullable = true)
 |-- Company: string (nullable = true)
 |-- Churn: integer (nullable = true)



In [5]:
data.select('Company').distinct().count()

873

In [6]:
by_comp = data.groupBy('Company').count()
by_comp.orderBy(by_comp['count'].desc()).show()

+--------------+-----+
|       Company|count|
+--------------+-----+
|Anderson Group|    4|
|    Wilson PLC|    3|
|  Williams PLC|    3|
|Perry and Sons|    2|
|     Gates Ltd|    2|
|    Walker Ltd|    2|
|   Davis Group|    2|
|      Webb PLC|    2|
|   Smith Group|    2|
|Smith and Sons|    2|
|     Ortiz Ltd|    2|
|     Smith Inc|    2|
|  Williams LLC|    2|
|      Rice PLC|    2|
|      Soto PLC|    2|
|    Nelson LLC|    2|
|      King LLC|    2|
|     Evans LLC|    2|
|Davis and Sons|    2|
|     Jones LLC|    2|
+--------------+-----+
only showing top 20 rows



As for the column 'Company', it is better to drop it than hold it. The amount of companies are 873, and each one are shown from 1 to 4 times.

In [7]:
data.select('Num_Sites').distinct().show()

+---------+
|Num_Sites|
+---------+
|      8.0|
|      7.0|
|      4.0|
|     11.0|
|     14.0|
|      3.0|
|     10.0|
|     13.0|
|      6.0|
|      5.0|
|      9.0|
|     12.0|
+---------+



In [9]:
data.select('Location').head(10)

[Row(Location='10265 Elizabeth Mission Barkerburgh, AK 89518'),
 Row(Location='6157 Frank Gardens Suite 019 Carloshaven, RI 17756'),
 Row(Location='1331 Keith Court Alyssahaven, DE 90114'),
 Row(Location='13120 Daniel Mount Angelabury, WY 30645-4695'),
 Row(Location='765 Tricia Row Karenshire, MH 71730'),
 Row(Location='6187 Olson Mountains East Vincentborough, PR 74359'),
 Row(Location='4846 Savannah Road West Justin, IA 87713-3460'),
 Row(Location='25271 Roy Expressway Suite 147 Brownport, FM 59852-6150'),
 Row(Location='3725 Caroline Stravenue South Christineview, MA 82059'),
 Row(Location='363 Sandra Lodge Suite 144 South Ann, WI 51655-7561')]

In [28]:
from pyspark.sql.functions import split, posexplode

In [27]:
data.select(split('Location', ' ')).head(10)[3][0][-1]

'30645-4695'

TypeError: object of type 'DataFrame' has no len()

In [36]:
for word in len(data):
    print(word)

TypeError: object of type 'DataFrame' has no len()